# Assignment 1: Wrangling and EDA
### Foundations of Machine Learning

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `airbnb_NYC.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the Minnesota police use of for data, `mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? For the remaining missing values, replace the `np.nan/None` values with the label `Missing`.
3. Dummy variable: For `metabric.csv`, convert the `Overall Survival Status` variable into a dummy/binary variable, taking the value 0 if the patient is deceased and 1 if they are living.
4. Missing values: For `airbnb_NYC.csv`, determine how many missing values of `Review Scores Rating` there are. Create a new variable, in which you impute the median score for non-missing observations to the missing ones. Why might this bias or otherwise negatively impact your results?

In [2]:
! git clone https://github.com/katiehutc/scratchpad
import urllib.request
import os
import zipfile
import os

def download_data(force=False):
    """Download and extract course data from Zenodo."""

    zip_path = 'data.zip'
    data_dir = './data'

    if not os.path.exists(zip_path) or force:
        print("Downloading course data...")
        urllib.request.urlretrieve(
            'https://zenodo.org/records/18235955/files/data.zip?download=1',
            zip_path
        )
        print("Download complete")

    if not os.path.exists(data_dir) or force:
        print("Extracting data files...")
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            zip_ref.extractall(data_dir)
        print("Data extracted")

    return data_dir


if __name__ == "__main__":
    download_data()


Cloning into 'scratchpad'...
remote: Enumerating objects: 33, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 33 (delta 7), reused 5 (delta 5), pack-reused 22 (from 1)
Receiving objects: 100% (33/33), 39.51 KiB | 674.00 KiB/s, done.
Resolving deltas: 100% (9/9), done.
Download complete
Extracting data files...
Data extracted


Part 1:

In [ ]:
import pandas as pd

df = pd.read_csv('/content/data/airbnb_NYC.csv', encoding="latin1")
# replaced commas with empty string
df["Price_clean"] = df["Price"].str.replace(",", "", regex=False)

# stripped the leading and trailing white spaces
df["Price_clean"] = df["Price_clean"].str.strip()

# converted string to number and changed empty strings to NaN
df["Price_clean"] = pd.to_numeric(df["Price_clean"], errors="coerce")

# count number of missing values
num_missing = df["Price_clean"].isna().sum()
print(df["Price"].str.contains(",", na=False).sum())
print("There are " + str(num_missing) + " missing values")


181
There are 0 missing values


Before formatting, 181 values had commas. After removing commas and converting the Price variable to numeric, there were no missing values. This indicates that all the numbers had valid price entries once formatting issues were resolved.

Part 2:

In [20]:
import pandas as pd

df = pd.read_csv("/content/data/mn_police_use_of_force.csv", encoding="latin1")

# inspect column
print(df["subject_injury"].value_counts(dropna=False))

# strip white spaces and use title case
df["subject_injury"] = df["subject_injury"].str.strip().str.title()

#set non yes/no values to Missing label
df.loc[
    ~df["subject_injury"].isin(["Yes", "No"]),
    "subject_injury"
] = "Missing"

# get proportion of missing values
prop_missing = (df["subject_injury"] == "Missing").mean()
print(f"{prop_missing * 100:.2f}% missing values")

#compare force with injury
pd.crosstab(
    df["force_type"],
    df["subject_injury"],
    dropna=False
)


subject_injury
NaN    9848
Yes    1631
No     1446
Name: count, dtype: int64
76.19% missing values


subject_injury               Missing    No   Yes
force_type                                      
Baton                              2     0     2
Bodily Force                    7051  1093  1286
Chemical Irritant               1421   131    41
Firearm                            0     2     0
Gun Point Display                 27    33    44
Improvised Weapon                 74    34    40
Less Lethal                       87     0     0
Less Lethal Projectile             0     1     2
Maximal Restraint Technique      170     0     0
Police K9 Bite                    31     2    44
Taser                            985   150   172

After cleaning the values in the subject_injury column, I found that 76.19% of the values were missing. This was done by calculating the number of missing values divided by the total values.

After that, I compared the subject_injury column with the force type column using the crosstab tool. Through this, you can see that for bodily force, chemical irritant, improvised weapon, less lethal, maximal restraint technique, police K9 bite, and taser, the missing values were much higher. These are almost all of the categories. This may indicate that police don't report subject injury as much when they use force.

Part 3:

In [27]:
import pandas as pd

df = pd.read_csv("/content/data/metabric.csv", encoding="latin1")
print(df["Overall Survival Status"].value_counts())

# convert to binary
df["Overall Survival Status"] = df["Overall Survival Status"].map({
    "1:DECEASED": 0,
    "0:LIVING": 1
})

print(df["Overall Survival Status"].value_counts())


Overall Survival Status
1:DECEASED    768
0:LIVING      575
Name: count, dtype: int64
Overall Survival Status
0    768
1    575
Name: count, dtype: int64


Here, I converted deceased patients to 0 and alive patients to 1 by mapping the variables.

Part 4:

In [38]:
import pandas as pd

df = pd.read_csv('/content/data/airbnb_NYC.csv', encoding="latin1")

# inspect column
# print(df["Review Scores Rating"].value_counts(dropna=False))

# count number missing
num_missing = df["Review Scores Rating"].isna().sum()
print(f"missing: {num_missing}")

# compute median
median = df["Review Scores Rating"].median()
print(f"median: {median}")

# impute
df["Review_Scores_Replace"] = df["Review Scores Rating"].fillna(median)
print(df["Review_Scores_Replace"].isna().sum())

# median after
median = df["Review Scores Rating"].median()
print(f"media after: {median}")

missing: 8323
median: 94.0
0
media after: 94.0


In this column, I found 8323 missing values. Then, I imputed by filling in the missing values with the median (94). Doing so can assume that the missing values were around the median value, and this is an incorrect assumption. There's no way to know what all the missing values, and this technique reduces the variability and doesn't reflect the true variation in the data. Additionally, if the values are missing for a reason for a purposeful reason, doing this technique can hide these reasons.

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks.

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work. What does work?
2. Drop any columns that do not contain data.
3. What is an observation? Carefully justify your answer, and explain how it affects your choices in cleaning and analyzing the data.
4. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
5. Clean the Age variable and make a histogram of the ages of the victims.
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?
7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.
8. Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Thoughts?

**Q3.** Open the "tidy_data.pdf" document available in `https://github.com/ds4e/wrangling`, which is a paper called *Tidy Data* by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?

**Q4.** This question looks at financial transfers from international actors to American universities. In particular, from which countries and giftors are the gifts coming from, and to which institutions are they going?

For this question, `.groupby([vars]).count()` and `.groupby([vars]).sum()` will be especially useful to tally the number of occurrences and sum the values of those occurrences.

1. Load the `ForeignGifts_edu.csv` dataset.
2. For `Foreign Gift Amount`, create a histogram and describe the variable. Describe your findings.
3. For `Gift Type`, create a histogram or value counts table. What proportion of the gifts are contracts, real estate, and monetary gifts?
4. What are the top 15 countries in terms of the number of gifts? What are the top 15 countries in terms of the amount given?
5. What are the top 15 institutions in terms of the total amount of money they receive? Make a histogram of the total amount received by all institutions.
6. Which giftors provide the most money, in total?

**Q5.** This question provides some practice doing exploratory data analysis and visualization.

We'll use the `college_completion.csv` dataset from the US Department of Education. The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `college_completion.csv` data with Pandas.
2. How many observations and variables are in the data? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see in words.
4. For `grad_100_value`, create a kernel density plot and describe table. Now condition on `control`, and produce a kernel density plot and describe tables for each type of institutional control. Which type of institution appear to have the most favorable graduation rates?
5. Make a scatterplot of `grad_100_value` by `aid_value`, and compute the covariance and correlation between the two variables. Describe what you see. Now make the same plot and statistics, but conditioning on `control`. Describe what you see. For which kinds of institutions does aid seem to vary positively with graduation rates?

**Q6.** In class, we talked about how to compute the sample mean of a variable $X$,
$$
m(X) = \dfrac{1}{N} \sum_{i=1}^N x_i
$$
and sample covariance of two variables $X$ and $Y$,
$$
\text{cov}(X,Y) = \dfrac{1}{N} \sum_{i=1}^N (x_i - m(X))(y_i - m(Y))).
$$
Recall, the sample variance of $X$ is
$$
s^2 = \dfrac{1}{N} \sum_{i=1}^N (x_i - m(X))^2.
$$
It can be very helpful to understand some basic properties of these statistics. If you want to write your calculations on a piece of paper, take a photo, and upload that to your GitHub repo, that's probably easiest.

We're going to look at **linear transformations** of $X$, $Y = a + bX$. So we take each value of $X$, $x_i$, and transform it as $y_i = a + b x_i$.

1. Show that $m(a + bX) = a+b \times m(X)$.
2. Show that $ \text{cov}(X,X) = s^2$.
3. Show that $\text{cov}(X,a+bY) = b \times \text{cov}(X,Y)$
4. Show that $\text{cov}(a+bX,a+bY) = b^2 \text{cov}(X,Y) $. Notice, this also means that $\text{cov}(bX, bX) = b^2 s^2$.
5. Suppose $b>0$ and let the median of $X$ be $\text{med}(X)$. Is it true that the median of $a+bX$ is equal to $a + b \times \text{med}(X)$? Is the IQR of $a + bX$ equal to $a + b \times \text{IQR}(X)$?
6. Show by example that the means of $X^2$ and $\sqrt{X}$ are generally not $(m(X))^2$ and $\sqrt{m(X)}$. So, the results we derived above really depend on the linearity of the transformation $Y = a + bX$, and transformations like $Y = X^2$ or $Y = \sqrt{X}$ will not behave in a similar way.

**Q7.** This question provides some practice doing exploratory data analysis and visualization.

We'll use the `ames_prices.csv` dataset. The "relevant" variables for this question are:
  - `price` - Sale price value of the house
  - `Bldg.Type` - Building type of the house (single family home, end-of-unit townhome, duplex, interior townhome, two-family conversion)

1. Load the `college_completion.csv` data with Pandas.
2. Make a kernel density plot of price and compute a describe table. Now, make a kernel density plot of price conditional on building type, and use `.groupby()` to make a describe type for each type of building. Which building types are the most expensive, on average? Which have the highest variance in transaction prices?
3. Make an ECDF plot of price, and compute the sample minimum, .25 quantile, median, .75 quantile, and sample maximum (i.e. a 5-number summary).
4. Make a boxplot of price. Are there outliers? Make a boxplot of price conditional on building type. What patterns do you see?
5. Make a dummy variable indicating that an observation is an outlier.
6. Winsorize the price variable, and compute a new kernel density plot and describe table. How do the results change?